vllm/distributed/kv_transfer/vllm_adapter.py

```diff
@@ -0,0 +1,444 @@
+"""vLLM distributed KV cache transfer API.
+These APIs are used in `vllm/worker/model_runner.py`.
+
+Currently supporting TP. The TP between prefill and decode instance needs to be 
+the same.
+
+Workflow (disaggregated prefill)
+- In prefill instance
+    - After prefill, vLLM `insert` its KV caches into a lookup buffer.
+    - The prefill instance will also open up a thread that listens to 
+      `drop_select` request.
+- In decode instance
+    - vLLM first runs `drop_select` to send input tokens and a mask on input 
+      tokens (we call it roi, region of interest) to prefill instance
+    - The prefill instance then respond to `drop_select` request by
+        - Finding a match in current lookup buffer.
+        - Clone and send the matched item out
+        - Delete the matched item in the lookup buffer to free up GPU memory.
+    - The decode vLLM then store the KV cache into paged memory.
+"""
+from typing import TYPE_CHECKING, List, Optional, Tuple, Union
+
+if TYPE_CHECKING:
+    from vllm.worker.model_runner import ModelInputForGPUWithSamplingMetadata
+
+from copy import deepcopy
+
+import torch
+from torch.distributed import Backend
+
+import vllm.distributed.kv_transfer.kv_lookup_buffer.simple_buffer as sklb
+import vllm.envs as envs
+from vllm import _custom_ops as ops
+from vllm.distributed.kv_transfer.kv_lookup_buffer.base import (
+    KVLookupBufferBase)
+from vllm.distributed.kv_transfer.kv_pipe.torch_distributed_pipe import (
+    TorchDistributedPipe)
+from vllm.logger import init_logger
+from vllm.sequence import IntermediateTensors
+
+logger = init_logger(__name__)
+
+# check VLLM_DISTRIBUTERD_KV_ROLE and set corresponding flags
+assert envs.VLLM_DISTRIBUTED_KV_ROLE in [None, "producer", "consumer", "both"],\
+    "VLLM_DISTRIBUTERD_KV_ROLE can only be producer, consumer or both."
+IS_DISTRIBUTED_KV_INSTANCE: bool = (envs.VLLM_DISTRIBUTED_KV_ROLE
+                                    in ["producer", "consumer", "both"])
+IS_KV_PRODUCER: bool = (envs.VLLM_DISTRIBUTED_KV_ROLE in ["producer", "both"])
+IS_KV_CONSUMER: bool = (envs.VLLM_DISTRIBUTED_KV_ROLE in ["consumer", "both"])
+
+# When the current instance is both KV producer and KV consumer,
+# it is likely connected to a KV storage service on CPU/disk
+# so the communication backend needs to be "gloo" for that case.
+DISTRIBUTED_BACKEND: str = "gloo" if (IS_KV_PRODUCER
+                                      and IS_KV_CONSUMER) else "nccl"
+# corresponding device
+DISTRIBUTED_DEVICE: str = "cpu" if (IS_KV_PRODUCER
+                                    and IS_KV_CONSUMER) else "cuda"
```

Comment:
Does it mean that it is not possible that the producer and consumer can run on the same GPU instance?

Reply:
The prefiller and the decoder can be placed on different GPUs within one node. But they cannot be placed on the same physical GPU card.